fairseq_cli/preprocess.py

```diff
@@ -45,6 +46,21 @@ def main(args):
 
     task = tasks.get_task(args.task)
 
+    def parse_source_args(args):
+        if args.srcdict:
+            if len(args.source_lang) != len(args.srcdict):
+                raise RuntimeError(
+                    "Number of source languages ({}) must match the number of source dictionaries ({}).".format(
+                        len(args.source_lang), len(args.srcdict)))
+            args.srcdict_uniq = list(set(args.srcdict))
```

Comment:
I don't think it messes up the order. I never overwrite `args.srcdict`. I store a set of unique dictionaries to `args.srcdict_uniq` (I admit I'd rather use a local variable instead of an `args` property) and create a `dict_index` as mapping from languages (in user-specified order) to the set of dictionaries. Then I load each dictionary only once on line 120 and map the dictionaries to languages using the `dict_index` on line 129.